### PR TITLE
Enhancement / Print “accessing as <user>” in client

### PR DESCRIPTION
**Pull request by EvgeniLeonti**
notion ticket: https://www.notion.so/dagshub/Print-accessing-as-user-in-client-92af32ddfce2477ba7184722634fb891?pvs=4

### CODE REVIEW

**Comment by EvgeniLeonti** [comment]
@kbolashev i have implemented initial version which works fine, but isn't efficient:
it will make another http request to get the username from the token.

in cases where we call `is_valid_token()`, this new http request can be skipped.

so we have 2 options, leave it as is, or make required adjustments to skip the redundant http call.
in order to achieve the 2nd option, i thought about:
1. adding username to `__token_cache`
2. maintain existing caches to support backwards compatibility (adding username to already existing tokens)

this will require more complicated logic, and i'm not sure the scope of this task covers that.

WDYT?

**Comment by kbolashev** [comment]
I think you can change the signature and name of `is_valid_token` to return the username and get 99% of the way there.
You'll only have to call it for the case of the EnvVar token and that's it.

File: dagshub/auth/tokens.py [hunk 1]
```diff
@@ -222,10 +231,23 @@ def is_valid_token(token: Union[str, Auth, DagshubTokenABC], host: str) -> bool:
             # 500's might be ok since they're server errors, so check only for 400's
             assert not (400 <= resp.status_code <= 499)
             if resp.status_code == 200:
```
Comment (kbolashev) [inline dagshub/auth/tokens.py:233]: A 500 will not return the dictionary, so probably should change it to this now:
```suggestion
            assert resp.status_code == 200
```

File: dagshub/auth/tokens.py [hunk 2]
```diff
@@ -222,10 +231,23 @@ def is_valid_token(token: Union[str, Auth, DagshubTokenABC], host: str) -> bool:
...
+                assert "login" in user
+                assert "username" in user
+                return user
         except AssertionError:
-            return False
+            return None
+
+    @staticmethod
+    def is_valid_token(token: Union[str, Auth, DagshubTokenABC], host: str) -> bool:
+        """
```
Comment (kbolashev) [inline dagshub/auth/tokens.py:243]: Great job on still leaving this! I honestly forgot it's used in a couple other places 🙃

File: dagshub/auth/tokens.py [hunk 3]
```diff
@@ -281,6 +303,24 @@ def _store_cache_file(self):
             logger.error(f"Error while storing DagsHub token cache: {traceback.format_exc()}")
             raise
 
+    def _print_accessing_as(self, user):
```
Comment (kbolashev) [inline dagshub/auth/tokens.py:306]:
```suggestion
    def _print_accessing_as(self, user: Dict):
```

File: dagshub/auth/tokens.py [hunk 4]
```diff
@@ -281,6 +303,24 @@ def _store_cache_file(self):
...
+        It called after successful token validation.
+
+        """
+        if self._accessing_as_was_printed:
+            return
+
+        username = user.get("username")
+        if username is None:
+            # todo handle error ? it should be defined
+            return
```
Comment (kbolashev) [inline dagshub/auth/tokens.py:319]: Something probably went SUPER wrong if we got the response from the API, but the structure of the dictionary is different

I say just skip the whole check and just do a regular dict access. If such a weird case happens, it's probably a backend error and there's no point in trying to handle it in the client
```suggestion
        username = user["username"]
```

File: tests/common/test_determine_repo.py [hunk 5]
```diff
@@ -2,13 +2,15 @@
...
 from typing import Generator, TypeVar
+import respx
 
 import pytest
 import pytest_git
 
 from dagshub.common.determine_repo import determine_repo, parse_dagshub_remote
 import dagshub.common.config
 from dagshub.common.errors import DagsHubRepoNotFoundError
+from tests.dda.test_tokens import valid_token_side_effect
```
Comment (kbolashev) [inline tests/common/test_determine_repo.py:13]: Maybe move this to `tests.util` or someplace like that.
Accessing across the test modules like that looks weird

**Review by kbolashev — CHANGES_REQUESTED** [review]
Looks great!

I left a bunch of minor things that need to be changed, but the logic is great.

File: dagshub/auth/tokens.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 import threading
 import traceback
 from typing import Optional, Dict, List, Set, Union
+from multiprocessing import AuthenticationError
```
Comment (kbolashev) [inline dagshub/auth/tokens.py:7]: Very weird module to import from, either create a new error class, or use RuntimeError

Examples: https://github.com/DagsHub/client/blob/master/dagshub/data_engine/model/errors.py

**Review by kbolashev — CHANGES_REQUESTED** [review]

File: dagshub/auth/tokens.py (hunk 6 above)
Comment (EvgeniLeonti) [reply dagshub/auth/tokens.py:7]: that was weird to me as well, took it from `dagshub/streaming/filesystem.py`

**Review by kbolashev — APPROVED** [review]
Congrats on the first feature PR for the client!!! 🥳

**Pull request merged**